rcl/src/rcl/arguments.c

```diff
@@ -0,0 +1,377 @@
+{
+  // assumes ASCII
+  return (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z') || (c >= '0' && c <= '9') || c == '_';
+}
+
+/// Parse an argument that may or may not be a remap rule
+/// \param[in] arg the argument to parse
+/// \param[in] allocator an allocator to use
+/// \param[in,out] output_rule input a zero intialized rule, output a fully initialized one
+/// \return RCL_RET_OK if a valid rule was parsed
+/// \return RLC_RET_ERROR if no valid rule was parsed
+/// \internal
+RCL_LOCAL
+rcl_ret_t
+_rcl_parse_remap_rule(
+  const char * arg,
+  rcl_allocator_t allocator,
+  rcl_remap_t * output_rule)
+{
+  size_t len_node_name = 0;
+  size_t len_match = 0;
+  size_t len_replacement = 0;
+
+  const char * separator = NULL;
+  const char * colon = NULL;
+  const char * match_begin = arg;
+  const char * replacement_begin = NULL;
+
+  // A valid rule has two parts separated by :=
+  separator = strstr(arg, ":=");
+  if (NULL == separator) {
+    return RCL_RET_ERROR;
+  }
+
+  replacement_begin = separator + 2;
+
+  // must have characters on both sides of the separator
+  len_match = separator - arg;
+  len_replacement = strlen(replacement_begin);
+  if (len_match <= 0 || len_replacement <= 0) {
+    return RCL_RET_ERROR;
+  }
+
+  colon = strchr(arg, ':');
+  if (NULL != colon) {
+    if (colon < separator) {
+      // If there is a : on the match side then there is a node-name prefix
+      match_begin = colon + 1;
+      len_node_name = colon - arg;
+      len_match = separator - match_begin;
+      // node name must have at least one character
+      if (len_node_name <= 0) {
+        return RCL_RET_ERROR;
+      }
+    } else if (colon > separator) {
+      // If the colon is on the replacement side then this couldn't be a valid rule
+      return RCL_RET_ERROR;
+    }
+  }
+
+  // match side must have at least 1 character
+  if (len_match <= 0) {
+    return RCL_RET_ERROR;
+  }
+
+  // Make sure node name contains only valid characters
+  for (size_t i = 0; i < len_node_name; ++i) {
+    if (!_rcl_valid_token_char(arg[i])) {
+      return RCL_RET_ERROR;
+    }
+  }
+
+  // Figure out what type of rule this is, default is to apply to topic and service names
+  rcl_remap_type_t type = RCL_TOPIC_REMAP | RCL_SERVICE_REMAP;
+  if (0 == strncmp("__ns", match_begin, len_match)) {
+    type = RCL_NAMESPACE_REMAP;
+  } else if (0 == strncmp("__node", match_begin, len_match)) {
+    type = RCL_NODENAME_REMAP;
+  }
+
+  if (type & (RCL_TOPIC_REMAP | RCL_SERVICE_REMAP)) {
+    // Replacement must be a valid topic name
+    char * copy_replacement = rcutils_strndup(replacement_begin, len_replacement, allocator);
+    if (NULL == copy_replacement) {
+      return RCL_RET_ERROR;
+    }
+    int validation_result;
+    size_t invalid_index;
+    rcl_ret_t ret = rcl_validate_topic_name(copy_replacement, &validation_result, &invalid_index);
+    if (ret != RCL_RET_OK || validation_result != RCL_TOPIC_NAME_VALID) {
+      ret = RCL_RET_ERROR;
```

Comment:
Better error messages using `validation_result` in 4f5180f61271532718679934328d3a213b04f7ef. `rcl_parse_arguments` logs the error messages as `DEBUG`. I logged them this way to avoid unnecessary console output when passing in non-ros arguments.